src/contracts/MigrationHelperV3V3.sol

```diff
@@ -0,0 +1,259 @@
+        );
+      }
+
+      (
+        RepayInput[] memory positionsToRepayWithAmounts,
+        address[] memory assetsToFlash,
+        uint256[] memory amountsToFlash,
+        uint256[] memory interestRatesToFlash
+      ) = _getFlashloanParams(positionsToRepay);
+
+      V3_TARGET_POOL.flashLoan(
+        address(this),
+        assetsToFlash,
+        amountsToFlash,
+        interestRatesToFlash,
+        msg.sender,
+        abi.encode(assetsToMigrate, positionsToRepayWithAmounts, msg.sender),
+        6671
+      );
+    }
+  }
+
+  /**
+   * @dev expected structure of the params:
+   *    assetsToMigrate - the list of supplied assets to migrate
+   *    positionsToRepay - the list of borrowed positions, asset address, amount and debt type should be provided
+   *    beneficiary - the user who requested the migration
+    @inheritdoc IMigrationHelper
+   */
+  function executeOperation(
+    address[] calldata,
+    uint256[] calldata,
+    uint256[] calldata,
+    address initiator,
+    bytes calldata params
+  ) external returns (bool) {
+    require(msg.sender == address(V3_TARGET_POOL), 'ONLY_V3_POOL_ALLOWED');
+    require(initiator == address(this), 'ONLY_INITIATED_BY_MIGRATION_HELPER');
+
+    (address[] memory assetsToMigrate, RepayInput[] memory positionsToRepay, address user) = abi
+      .decode(params, (address[], RepayInput[], address));
+
+    for (uint256 i = 0; i < positionsToRepay.length; i++) {
+      V3_SOURCE_POOL.repay(
+        positionsToRepay[i].asset,
+        positionsToRepay[i].amount,
+        positionsToRepay[i].rateMode,
+        user
+      );
+    }
+
+    _migrationNoBorrow(user, assetsToMigrate);
+
+    return true;
+  }
+
+  /// @inheritdoc IMigrationHelper
+  function getMigrationSupply(
+    address asset,
+    uint256 amount
+  ) external view virtual returns (address, uint256) {
+    return (asset, amount);
+  }
+
+  // helper method to get v3 reserves addresses for migration
+  // mostly needed to make overrides simpler on specific markets with many available reserves, but few valid
+  function _getV3Reserves() internal virtual returns (address[] memory) {
+    return V3_SOURCE_POOL.getReservesList();
+  }
+
+  function _migrationNoBorrow(address user, address[] memory assets) internal {
+    address asset;
+    IERC20WithPermit aToken;
+    uint256 aTokenAmountToMigrate;
+    uint256 aTokenBalanceAfterReceiving;
+
+    for (uint256 i = 0; i < assets.length; i++) {
+      asset = assets[i];
+      aToken = aTokens[asset];
+
+      require(asset != address(0) && address(aToken) != address(0), 'INVALID_OR_NOT_CACHED_ASSET');
+
+      aTokenAmountToMigrate = aToken.balanceOf(user);
+      aToken.safeTransferFrom(user, address(this), aTokenAmountToMigrate);
+
+      // this part of logic needed because of the possible 1-3 wei imprecision after aToken transfer, for example on stETH
+      aTokenBalanceAfterReceiving = aToken.balanceOf(address(this));
+      if (
```

Comment:
what's the reasoning here?
why not checking balances before and after the transfer?